### DISCUSSION TITLE
Lazy open

### DISCUSSION THREAD

**Discussion by nstinus** [GitoxideLabs/gitoxide#2001] [Q&A]
Hi,

Given a commit sha, I need to get the committer and author name and email. Access the content of a bunch of files and get the list of files touched between a ref and this commit.

I recently replaced my git subprocesses calls with gitoxide in a pretty large repo and unfortunately, the performance is pretty bad. I've gathered an estimate of the times here:

| Operation | Subprocess | Gix|
|-|-|-|
| Metadata | 50ms | 200ms*|
| Cat file | 50ms| <5ms |
| Diff-tree| 50ms|>80ms |

* 180ms for `ThreadSafeRepository::open()` followed by 20ms for `repo.find_commit(repo.rev_parse_single(rev)?.detach())?`

I tried accessing the metadata using `gix cat` and the perf is consistent with what I see in my app (~200ms).

I'd like to be able to approach the performance I see with the subprocess. I suspect gix loads a lot more from the repo when it is opened that is strictly required to get the info I need vs git that dereferences a few things only.

Is my assessment correct? Is there a way to access info in a repo in a more lazy way? Is there anything you recommend in my case? Thank you for your help!

**Comment by Byron** [comment]
Thanks so much for reporting - I'd really appreciate an issue for this.

`gix::open()` and related calls are already lazy, but since testing this isn't straightforward it may have regressed.
The last time I heard about this was when StackedGit reported that changing their test-suite from `git2` to `gix` made it several times faster due to vastly improved startup times.

Also, after the integration with `helix` there were issues around how many `stat` calls it performs during startup on very slow (network) drives. After the fix it did only what it had to do and performance was greatly improved.

If you choose to create an issue, there I would ask if you have a way to do an instrumented run to see what exactly is slow. I suspect that an `strace` run would already be very revealing.

Since `cat file` is fast, it may also be a misunderstanding and it's the operation (like `diff tree`) that is slow. There are performance notes on these functions, stating that an object cache has to be configured to speed up diffs significantly.

I hope that helps.
